frontend/components/feature/QueryFilters/MultiSelect.tsx

```diff
@@ -1,43 +1,149 @@
-import { Button } from "@/components/base/button";
-import { ActionButton } from "../../base/Buttons";
+import { useCallback, useEffect, useState } from "react";
+import { Button } from "@/components/base/ui/button";
+import { ActionButton } from "@/components/base/Buttons";
+import { Aggregation } from "@/api/api";
 
-interface IProps {}
+interface IProps {
+  data: any[];
+  appliedItems?: string[];
+  maxVisibleItems?: number;
+  searchVisible?: boolean;
+}
+
+function searchOptions(search: string, data: any[]) {
+  return data.filter((option) => option.key.includes(search));
+}
+
+export function MultiSelect({
+  data,
+  maxVisibleItems = 10,
+  searchVisible = false,
+  appliedItems = [],
+}: IProps) {
+  const [items, setItems] = useState(data);
+  const [appliedSelectedItems, setAppliedSelectedItems] =
+    useState<string[]>(appliedItems);
+  const [selectedItems, setSelectedItems] = useState<string[]>([]);
+  const [visibleItemsCount, setVisibleItemsCount] = useState(
+    maxVisibleItems < items.length ? maxVisibleItems : items.length,
+  );
+  const [hasUnappliedItems, setHasUnappliedItems] = useState(false);
+
+  // Memoize these functions with useCallback
+  const showMore = () => {
+    let count = visibleItemsCount + maxVisibleItems;
+    setHasUnappliedItems(true);
+    setVisibleItemsCount(count > items.length ? items.length : count);
+  };
+
+  const selectAll = useCallback(() => {
+    if (selectedItems.length === items.length) {
+      return;
+    }
+    setHasUnappliedItems(true);
+    setSelectedItems(items.map((f) => f.key));
+  }, [items]);
+
+  const unSelectAll = useCallback(() => {
+    if (selectedItems.length === 0) {
+      return;
+    }
+    setHasUnappliedItems(true);
+    setSelectedItems([]);
+  }, []);
+
+  const itemSelected = useCallback((item: Aggregation) => {
+    let newList: string[] = [];
+    if (selectedItems.some((f) => f === item.key)) {
+      newList = selectedItems.filter((f) => f !== item.key);
+    } else {
+      newList = [...selectedItems, item.key!];
+    }
+    setHasUnappliedItems(true);
+    setSelectedItems(newList);
+  }, []);
+
+  const reset = useCallback(() => {
+    setHasUnappliedItems(false);
+    setSelectedItems([...appliedSelectedItems]);
+  }, [appliedSelectedItems]);
+
+  const apply = useCallback(() => {
+    setHasUnappliedItems(false);
+    setAppliedSelectedItems(selectedItems);
+  }, [selectedItems]);
 
-export function MultiSelect({}: IProps) {
   return (
     <div className="bg-white p-2 w-full max-w-md">
-      <input
-        type="text"
-        placeholder="Search ..."
-        className="w-full p-2 mb-4 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500"
-      />
+      {searchVisible && (
+        <input
```

Comment:
Maybe use the Input component from `base/ui`

Reply:
fixed